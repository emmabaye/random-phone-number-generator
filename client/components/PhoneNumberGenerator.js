/* eslint-disable react/sort-comp */
import React, { Component } from "react";
import axios from "axios";


export default class PhoneNumberGenerator extends Component {
    state = {
      phoneNumbers: [],
      ascending: true,
      loading: false,
    };

    generateNewNumbers = (url) => {
      this.setState({
        loading: true,
      });
      axios.get(url)
        .then((response) => {
          console.log(response);
          this.setState({
            phoneNumbers: response.data.sort(),
            ascending: true,
            loading: false
          });
        })
        .catch((error) => {
          console.log("error ", error);
        });
    };

    toggleOrder = () => {
      this.setState({
        phoneNumbers: this.state.phoneNumbers.reverse(),
        ascending: !this.state.ascending
      });
    }

    componentDidMount() {
      this.generateNewNumbers("/download/generatedPhoneNumbers.json");
    }

    render() {
      return (
        <section id="app-section">
          <h2>Random Phone Number Generator</h2>
          <button
            className="btn btn-primary"
            onClick={() => this.generateNewNumbers('/api/generate-numbers')}>
          Generate New Numbers
          </button>
          <a
            className="btn btn-primary"
            href="/download/generatedPhoneNumbers.json"
            target="_blank" >
          Download As JSON
          </a>
          <a className="btn btn-primary"
            href="/download/generatedPhoneNumbers.txt"
            target="_blank">
          Download As Text
          </a>
          <a
            className="btn btn-primary"
            href="/download/generatedPhoneNumbers.csv"
            target="_blank">Download As CSV
          </a>
          <h5>Total Phone Numbers Generated: {
            !this.state.loading ? this.state.phoneNumbers.length : "Loading..."
          }</h5>
          <table className="table">
            <thead>
              <tr>
                <th scope="col">#</th>
                <th
                  scope="col"
                  className="toggleHeader"
                  onClick={() => this.toggleOrder()}>
                Phone Number{
                    this.state.ascending ? <i className="fa fa-sort-up" /> : <i className="fa fa-sort-down" />
                  }
                </th>
              </tr>
            </thead>
            <tbody>{
              this.state.phoneNumbers.map((phoneNumber, index) => (<tr key={index}>
                <th scope="row">{
                  this.state.ascending ? index + 1 : this.state.phoneNumbers.length - index
                }
                </th>
                <td>{phoneNumber}</td>
              </tr>))
            }
            </tbody>
          </table>
        </section>
      );
    }
}
