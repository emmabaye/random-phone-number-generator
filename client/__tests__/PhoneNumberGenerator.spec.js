/* eslint-disable import/no-extraneous-dependencies */
import PhoneNumberGenerator from '../components/PhoneNumberGenerator';

describe("PhoneNumberGenerator component ", () => {
  const mockData = ["0123456789", "0987654321"];

  beforeEach(() => {
    axiosMock
      .onAny("/download/generatedphonenumbers.json")
      .reply(200, mockData);
    axiosMock
      .onAny("/api/generate-numbers")
      .reply(200, mockData);
  });

  afterEach(() => axiosMock.reset());

  const wrapper = shallow(<PhoneNumberGenerator />);

  it("should render correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should handle on click for generating new numbers", () => {
    wrapper.setState({
      phoneNumbers: [],
      loading: false,
      ascending: false,
    });

    const generateNumbersButton = wrapper.find('.btn').at(0);

    generateNumbersButton.simulate("click");

    expect(wrapper.state().loading).toEqual(true);
  });

  it("should handle toggle order ", () => {
    const toggleHeader = wrapper.find('.toggleHeader').at(0);
    toggleHeader.simulate("click");

    expect(wrapper.state().ascending).toBe(false);
  });
});
