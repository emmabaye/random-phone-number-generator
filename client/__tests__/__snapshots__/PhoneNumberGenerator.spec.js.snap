// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhoneNumberGenerator component  should render correctly 1`] = `
<section
  id="app-section"
>
  <h2>
    Random Phone Number Generator
  </h2>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Generate New Numbers
  </button>
  <a
    className="btn btn-primary"
    href="/download/generatedphonenumbers.json"
    target="_blank"
  >
    Download As JSON
  </a>
  <a
    className="btn btn-primary"
    href="/download/generatedphonenumbers.txt"
    target="_blank"
  >
    Download As Text
  </a>
  <a
    className="btn btn-primary"
    href="/download/generatedphonenumbers.csv"
    target="_blank"
  >
    Download As CSV
  </a>
  <h5>
    Total Phone Numbers Generated: 
    Loading...
  </h5>
  <table
    className="table"
  >
    <thead>
      <tr>
        <th
          scope="col"
        >
          #
        </th>
        <th
          className="toggleHeader"
          onClick={[Function]}
          scope="col"
        >
          Phone Number
          <i
            className="fa fa-sort-up"
          />
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</section>
`;
